### PR TITLE
Feat/selected filters component

### PR DESCRIPTION
**Pull request by mohdashraf010897**
**PR Type** `Feature`

**Description**  A `<SelectedFilters/>` UI component for Flutter.

[📔  Notion](https://www.notion.so/appbase/SelectedFilters-component-Flutter-d00135ce593441d8ba5ee666caec1ee2)

[🎥  Loom ](https://www.loom.com/share/e8e03bd7c00143f5abe359b8ae5b79fa)

### CODE REVIEW

**Review by bietkul — CHANGES_REQUESTED** [review]
## Functionality Issues

- Only display value when present. Why are you displaying the `map` value?

- Why do `range-input` not update its state once filter gets cleared? Check at 1:14. There must a single source of truth for `value` which is `SearchController`. You should avoid maintaining local state. 

- It is clearly mentioned on Notion to use `Join` for array fields. Why do you render it as an array?

- `ClearAll` and `onClear` actions are throwing the error. It seems some invalid query is getting triggered.

- Assert when `defaultValues` are not present and `resetToDefault` or `hideDefaultValues` are set.

File: flutter_searchbox_ui/README.md [hunk 1]
```diff
@@ -220,73 +221,138 @@ class FlutterSearchBoxUIApp extends StatelessWidget {
...
             backgroundColor: Colors.white.withOpacity(.9),
           ),
-          body: Center(
-            // A custom UI widget to render earthquakes markers
-            child: ReactiveGoogleMap(
-              id: 'map-widget',
-              // To update markers when magnitude gets changed
-              react: const {
-                "and": "range-selector",
+          // floatingActionButton: SelectedFilters(),
```
Comment (bietkul) [inline flutter_searchbox_ui/README.md]:
```suggestion
```

File: flutter_searchbox_ui/README.md [hunk 2]
```diff
@@ -220,73 +221,138 @@ class FlutterSearchBoxUIApp extends StatelessWidget {
...
+          bottomNavigationBar: Padding(
+            padding: EdgeInsets.all(20.0),
+            // SelectedFilters: a widget to track all active filters
+            child: SelectedFilters(
+              subscribeTo: const ['range-selector', 'map-widget'],
```
Comment (bietkul) [inline flutter_searchbox_ui/README.md]:
```suggestion
              subscribeTo: const ['range-selector'],
```

File: flutter_searchbox_ui/example/lib/main.dart [hunk 3]
```diff
@@ -187,90 +187,138 @@ class FlutterSearchBoxUIApp extends StatelessWidget {
...
             backgroundColor: Colors.white.withOpacity(.9),
           ),
-          body: Center(
-            // A custom UI widget to render earthquakes markers
-            child: ReactiveGoogleMap(
-              id: 'map-widget',
-              // To update markers when magnitude gets changed
-              react: const {
-                "and": "range-selector",
+          // floatingActionButton: SelectedFilters(),
```
Comment (bietkul) [inline flutter_searchbox_ui/example/lib/main.dart]:
```suggestion
```

File: flutter_searchbox_ui/example/lib/main.dart [hunk 4]
```diff
@@ -187,90 +187,138 @@ class FlutterSearchBoxUIApp extends StatelessWidget {
...
+          bottomNavigationBar: Padding(
+            padding: EdgeInsets.all(20.0),
+            // SelectedFilters: a widget to track all active filters
+            child: SelectedFilters(
+              subscribeTo: const ['range-selector', 'map-widget'],
```
Comment (bietkul) [inline flutter_searchbox_ui/example/lib/main.dart]:
```suggestion
              subscribeTo: const ['range-selector'],
```

File: flutter_searchbox_ui/lib/src/components/range_input.dart [hunk 5]
```diff
@@ -1016,6 +1016,21 @@ class _RangeInputInnerState extends State<RangeInputInner> {
   @override
   void initState() {
     super.initState();
+    widget.searchController.subscribeToStateChanges((changes) {
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/components/range_input.dart]: @mohdashraf010897 You don't need to subscribe here. You should get the controller `value` property from `SearchWidgetConnector`.

File: flutter_searchbox_ui/lib/src/components/range_input.dart [hunk 6]
```diff
@@ -1305,6 +1320,31 @@ class _DropdownState extends State<Dropdown> {
     }
   }
 
+  @override
+  void didUpdateWidget(Dropdown oldWidget) {
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/components/range_input.dart:1355]: Revisit it based on the above comment.

File: flutter_searchbox_ui/lib/src/components/selected_filters.dart [hunk 7]
```diff
@@ -0,0 +1,363 @@
+import 'package:flutter/material.dart';
+import 'package:searchbase/searchbase.dart';
+import 'package:flutter_searchbox/flutter_searchbox.dart';
+import '../utils.dart';
+
+class BuildOptions {
+  /// A map of search controller id to controller value
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/components/selected_filters.dart]:
```suggestion
  /// A map of search controller id to [SearchController]
```

File: flutter_searchbox_ui/lib/src/components/selected_filters.dart [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+    required this.clearValues,
+  });
+}
+
+// It creates a selectable filter UI view displaying the current selected values from other active widgets.
+//
+// This component is useful for improving selection accessibility of other components.
+//
+// Examples Use(s):
+//    - displaying all the user selected facet filters together in the main view area for better accessibility.
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/components/selected_filters.dart]:
```suggestion
/// It creates a selectable filter UI view displaying the current selected values from other active widgets.
///
/// This component is useful for improving selection accessibility of other components.
///
/// Examples Use(s):
///    - displaying all the user selected facet filters together in the main view area for better accessibility.
```

File: flutter_searchbox_ui/lib/src/components/selected_filters.dart [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+class SelectedFilters extends StatefulWidget {
+  /// A list of component ids to subscribe to.
+  /// [Optional]
+  /// This property allows users to select the search contollerer for which
+  /// they want to display the filter value.
+  /// For example,
+  /// ```dart
+  /// SelectedFilters(
+  ///   ...
+  ///   subscribeTo: const ['range-selector', 'map-widget'],
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/components/selected_filters.dart]:
```suggestion
  ///   subscribeTo: const ['range-selector'],
```

File: flutter_searchbox_ui/lib/src/components/selected_filters.dart [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+  ///      spacing: 16.0,
+  ///     crossAxisAlignment: WrapCrossAlignment.start,
+  ///     children: widgets,
+  ///    );
+  ///   },
+  ///   ...
+  /// )
+  /// ```
+
+  final Widget Function([BuildOptions? options])? buildFilters;
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/components/selected_filters.dart]: `options` has to be used, so shouldn't be an optional param.

File: flutter_searchbox_ui/lib/src/components/selected_filters.dart [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+    final activeWidgets = this.activeWidgets;
+    if (widget.onClear != null) {
+      widget.onClear!(id, activeWidgets[id]?.value);
+    }
+
+    _selectedFilters.remove(id);
+    activeWidgets[id]?.setValue(resetTo ?? getResetValue(id),
+        options: Options(
+          triggerCustomQuery: true,
+          triggerDefaultQuery: true,
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/components/selected_filters.dart]: Is there a reason to trigger the default query?

File: flutter_searchbox_ui/lib/src/utils.dart [hunk 12]
```diff
@@ -66,3 +67,59 @@ Location? getLocationObject(dynamic location) {
...
+  if (value is String) {
+    return value;
+  } else if (value is num) {
+    return value.toString();
+  } else if (value is List) {
+    return value.join(", ");
+  } else if (value is Map) {
+    return processFilterValues(prepareValueList(value));
+  } else {
+    return value.toString();
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/utils.dart:98]: @mohdashraf010897 Do you handle empty string? It shouldn't display empty string as a selected filter.

File: flutter_searchbox_ui/lib/src/utils.dart [hunk 13]
```diff
@@ -66,3 +67,59 @@ Location? getLocationObject(dynamic location) {
...
+  }
+}
+
+bool isEqual(dynamic value, dynamic defaultValue) {
```
Comment (bietkul) [inline flutter_searchbox_ui/lib/src/utils.dart:102]: @mohdashraf010897 Can you use `isEqual` method defined in `searchbase`?
Btw why are you using it?

**Review by bietkul — CHANGES_REQUESTED** [review]

File: flutter_searchbox_ui/lib/src/utils.dart (hunk 12 above)
Comment (mohdashraf010897) [reply flutter_searchbox_ui/lib/src/utils.dart:98]: yes 
` if (value == null || value.isEmpty) {
    return "";
  }`

File: flutter_searchbox_ui/lib/src/components/selected_filters.dart (hunk 11 above)
Comment (mohdashraf010897) [reply flutter_searchbox_ui/lib/src/components/selected_filters.dart]: solved

File: flutter_searchbox_ui/lib/src/components/selected_filters.dart (hunk 10 above)
Comment (mohdashraf010897) [reply flutter_searchbox_ui/lib/src/components/selected_filters.dart]: fixed

Comment (mohdashraf010897) [reply flutter_searchbox_ui/lib/src/components/selected_filters.dart]: `final Widget Function(BuildOptions options)? buildFilters;`

File: flutter_searchbox_ui/lib/src/components/range_input.dart (hunk 5 above)
Comment (mohdashraf010897) [reply flutter_searchbox_ui/lib/src/components/range_input.dart]: used below

> @override
>   void didUpdateWidget(RangeInputInner oldWidget) {
>     super.didUpdateWidget(oldWidget);
>     var updatedValue = widget.searchController.value;
>     ...
>   }

File: flutter_searchbox_ui/lib/src/utils.dart (hunk 13 above)
Comment (mohdashraf010897) [reply flutter_searchbox_ui/lib/src/utils.dart:102]: deep equality purpose

File: flutter_searchbox_ui/example/lib/main.dart [hunk 14]
```diff
@@ -83,12 +88,14 @@ class FlutterSearchBoxUIApp extends StatelessWidget {
...
-                  return Future.error(value);
+                if (value is Map<String, dynamic>) {
+                  final Map<String, dynamic> mapValue =
+                      value as Map<String, dynamic>;
+                  if (mapValue['start'] == 0 && mapValue['end'] == null) {
+                    return Future.error(value);
+                  }
                 }
-
+                print('beforeValueChange $value');
```
Comment (bietkul) [inline flutter_searchbox_ui/example/lib/main.dart]:
```suggestion
```

**Review by bietkul — CHANGES_REQUESTED** [review]

**Pull request merged**